File: scripts/generate-icons.py
#!/usr/bin/env python3
"""Generate rounded, cute app icons for 小星星成长宝"""

import math
from PIL import Image, ImageDraw, ImageFilter, ImageFont

def rounded_star_points(cx, cy, outer_r, inner_r, points=5, round_factor=0.15):
    """Generate a rounded star path as a list of (x, y) points with extra interpolation for smoothness."""
    raw_points = []
    for i in range(points * 2):
        angle = math.radians(-90 + i * 360 / (points * 2))
        r = outer_r if i % 2 == 0 else inner_r
        raw_points.append((cx + r * math.cos(angle), cy + r * math.sin(angle)))

    # Subdivide with bezier-like smoothing
    smoothed = []
    n = len(raw_points)
    subdivisions = 8
    for i in range(n):
        p0 = raw_points[(i - 1) % n]
        p1 = raw_points[i]
        p2 = raw_points[(i + 1) % n]
        p3 = raw_points[(i + 2) % n]
        for t_step in range(subdivisions):
            t = t_step / subdivisions
            # Catmull-Rom spline
            t2 = t * t
            t3 = t2 * t
            x = 0.5 * ((2 * p1[0]) +
                       (-p0[0] + p2[0]) * t +
                       (2*p0[0] - 5*p1[0] + 4*p2[0] - p3[0]) * t2 +
                       (-p0[0] + 3*p1[0] - 3*p2[0] + p3[0]) * t3)
            y = 0.5 * ((2 * p1[1]) +
                       (-p0[1] + p2[1]) * t +
                       (2*p0[1] - 5*p1[1] + 4*p2[1] - p3[1]) * t2 +
                       (-p0[1] + 3*p1[1] - 3*p2[1] + p3[1]) * t3)
            smoothed.append((x, y))

    # Blend toward circle for extra roundness
    result = []
    avg_r = (outer_r + inner_r) / 2
    for (x, y) in smoothed:
        dx, dy = x - cx, y - cy
        dist = math.sqrt(dx*dx + dy*dy)
        if dist > 0:
            circle_x = cx + avg_r * dx / dist
            circle_y = cy + avg_r * dy / dist
            x = x * (1 - round_factor) + circle_x * round_factor
            y = y * (1 - round_factor) + circle_y * round_factor
        result.append((x, y))

    return result


def draw_rounded_rect(draw, bbox, radius, fill):
    """Draw a rounded rectangle."""
    x0, y0, x1, y1 = bbox
    r = radius
    # Main rectangles
    draw.rectangle([x0 + r, y0, x1 - r, y1], fill=fill)
    draw.rectangle([x0, y0 + r, x1, y1 - r], fill=fill)
    # Corners
    draw.pieslice([x0, y0, x0 + 2*r, y0 + 2*r], 180, 270, fill=fill)
    draw.pieslice([x1 - 2*r, y0, x1, y0 + 2*r], 270, 360, fill=fill)
    draw.pieslice([x0, y1 - 2*r, x0 + 2*r, y1], 90, 180, fill=fill)
    draw.pieslice([x1 - 2*r, y1 - 2*r, x1, y1], 0, 90, fill=fill)


def create_icon(size, output_path, is_maskable=False):
    """Create a cute rounded star icon at the given size."""
    # Use 4x supersampling for smooth edges
    ss = 4
    s = size * ss
    img = Image.new('RGBA', (s, s), (0, 0, 0, 0))
    draw = ImageDraw.Draw(img)

    # Background - warm gradient-like effect
    # Main background with rounded corners
    bg_color_top = (255, 192, 30)      # Warm golden yellow
    bg_color_bottom = (255, 165, 0)    # Deeper orange-gold

    if is_maskable:
        # Maskable icons need full bleed background
        corner_r = int(s * 0.05)
        draw_rounded_rect(draw, [0, 0, s, s], corner_r, bg_color_top)
    else:
        # Standard icon with nice rounded corners
        corner_r = int(s * 0.22)
        draw_rounded_rect(draw, [0, 0, s, s], corner_r, bg_color_top)

    # Subtle gradient overlay (top lighter, bottom deeper)
    for y in range(s):
        alpha = int(40 * (y / s))
        overlay_color = (200, 120, 0, alpha)
        draw.line([(0, y), (s, y)], fill=overlay_color)

    # Star parameters
    cx, cy = s // 2, int(s * 0.46)  # Slightly above center
    if is_maskable:
        outer_r = int(s * 0.28)
    else:
        outer_r = int(s * 0.32)
    inner_r = int(outer_r * 0.45)  # Chubbier inner ratio for cuter look

    # Draw soft shadow for the star
    shadow_offset = int(s * 0.015)
    shadow_points = rounded_star_points(cx, cy + shadow_offset, outer_r + int(s*0.01), inner_r + int(s*0.01), round_factor=0.2)
    shadow_img = Image.new('RGBA', (s, s), (0, 0, 0, 0))
    shadow_draw = ImageDraw.Draw(shadow_img)
    shadow_draw.polygon(shadow_points, fill=(180, 120, 0, 60))
    shadow_img = shadow_img.filter(ImageFilter.GaussianBlur(radius=int(s * 0.02)))
    img = Image.alpha_composite(img, shadow_img)
    draw = ImageDraw.Draw(img)

    # Draw the main rounded star (white with slight cream tint)
    star_points = rounded_star_points(cx, cy, outer_r, inner_r, round_factor=0.2)
    draw.polygon(star_points, fill=(255, 255, 255))

    # Draw cute face on the star
    face_cy = cy + int(outer_r * 0.05)

    # Eyes - cute round dots
    eye_r = int(outer_r * 0.08)
    eye_spacing = int(outer_r * 0.25)
    eye_y = face_cy - int(outer_r * 0.05)

    # Left eye
    draw.ellipse([
        cx - eye_spacing - eye_r, eye_y - eye_r,
        cx - eye_spacing + eye_r, eye_y + eye_r
    ], fill=(80, 60, 40))

    # Right eye
    draw.ellipse([
        cx + eye_spacing - eye_r, eye_y - eye_r,
        cx + eye_spacing + eye_r, eye_y + eye_r
    ], fill=(80, 60, 40))

    # Eye highlights (small white dots)
    highlight_r = int(eye_r * 0.45)
    highlight_offset = int(eye_r * 0.3)
    draw.ellipse([
        cx - eye_spacing - highlight_offset - highlight_r, eye_y - highlight_offset - highlight_r,
        cx - eye_spacing - highlight_offset + highlight_r, eye_y - highlight_offset + highlight_r
    ], fill=(255, 255, 255))
    draw.ellipse([
        cx + eye_spacing - highlight_offset - highlight_r, eye_y - highlight_offset - highlight_r,
        cx + eye_spacing - highlight_offset + highlight_r, eye_y - highlight_offset + highlight_r
    ], fill=(255, 255, 255))

    # Cute blush cheeks (soft pink circles)
    blush_r = int(outer_r * 0.1)
    blush_y = face_cy + int(outer_r * 0.08)
    blush_spacing = int(outer_r * 0.38)

    blush_img = Image.new('RGBA', (s, s), (0, 0, 0, 0))
    blush_draw = ImageDraw.Draw(blush_img)
    blush_draw.ellipse([
        cx - blush_spacing - blush_r, blush_y - blush_r,
        cx - blush_spacing + blush_r, blush_y + blush_r
    ], fill=(255, 150, 150, 100))
    blush_draw.ellipse([
        cx + blush_spacing - blush_r, blush_y - blush_r,
        cx + blush_spacing + blush_r, blush_y + blush_r
    ], fill=(255, 150, 150, 100))
    blush_img = blush_img.filter(ImageFilter.GaussianBlur(radius=int(s * 0.008)))
    img = Image.alpha_composite(img, blush_img)
    draw = ImageDraw.Draw(img)

    # Cute smile (small arc)
    smile_width = int(outer_r * 0.22)
    smile_y = face_cy + int(outer_r * 0.1)
    smile_thickness = max(int(s * 0.006), 2)
    draw.arc([
        cx - smile_width, smile_y - int(smile_width * 0.5),
        cx + smile_width, smile_y + int(smile_width * 0.8)
    ], 10, 170, fill=(80, 60, 40), width=smile_thickness)

    # Small sparkles around the star for extra cuteness
    sparkle_positions = [
        (cx + int(s * 0.28), cy - int(s * 0.22), int(s * 0.025)),
        (cx - int(s * 0.30), cy - int(s * 0.18), int(s * 0.018)),
        (cx + int(s * 0.18), cy + int(s * 0.28), int(s * 0.015)),
        (cx - int(s * 0.25), cy + int(s * 0.25), int(s * 0.02)),
    ]

    for sx, sy, sr in sparkle_positions:
        # Four-point sparkle
        draw.ellipse([sx - sr, sy - sr//3, sx + sr, sy + sr//3], fill=(255, 255, 255, 200))
        draw.ellipse([sx - sr//3, sy - sr, sx + sr//3, sy + sr], fill=(255, 255, 255, 200))

    # Add app name text at bottom (small, cute)
    text_y = cy + int(s * 0.30)
    text_size = int(s * 0.055)

    # Simple text rendering
    try:
        # Try system fonts
        for font_path in [
            '/System/Library/Fonts/PingFang.ttc',
            '/System/Library/Fonts/STHeiti Light.ttc',
            '/System/Library/Fonts/Hiragino Sans GB.ttc',
        ]:
            try:
                font = ImageFont.truetype(font_path, text_size)
                break
            except:
                continue
        else:
            font = ImageFont.load_default()
    except:
        font = ImageFont.load_default()

    text = "成长宝"
    bbox_text = draw.textbbox((0, 0), text, font=font)
    tw = bbox_text[2] - bbox_text[0]
    # Draw text shadow
    draw.text((cx - tw//2 + 1, text_y + 1), text, fill=(180, 120, 0, 120), font=font)
    # Draw text
    draw.text((cx - tw//2, text_y), text, fill=(255, 255, 255), font=font)

    # Downsample with high-quality resampling
    img = img.resize((size, size), Image.LANCZOS)

    # Convert to RGB for PNG (no transparency for app icons)
    final = Image.new('RGB', (size, size), (255, 192, 30))
    final.paste(img, mask=img.split()[3])

    final.save(output_path, 'PNG', quality=95)
    print(f"Generated: {output_path} ({size}x{size})")


if __name__ == '__main__':
    base_path = '/Users/Muna/Documents/GitHub/vanvan/public'

    # Generate all required sizes
    create_icon(180, f'{base_path}/apple-touch-icon-180x180.png')
    create_icon(192, f'{base_path}/pwa-192x192.png')
    create_icon(512, f'{base_path}/pwa-512x512.png')
    create_icon(512, f'{base_path}/maskable-icon-512x512.png', is_maskable=True)

    print("\nAll icons generated successfully!")

File: scripts/generate-icons.mjs
import sharp from 'sharp'

function createIconSvg(size, maskable = false) {
  const s = size
  const cx = s / 2
  const cy = s * 0.46
  const cornerR = maskable ? s * 0.02 : s * 0.22

  // Star geometry - plump, rounded
  const outerR = s * 0.30
  const innerR = outerR * 0.46

  // Generate smooth rounded star path
  const rawPts = []
  for (let i = 0; i < 10; i++) {
    const angle = -Math.PI / 2 + i * Math.PI * 2 / 10
    const r = i % 2 === 0 ? outerR : innerR
    rawPts.push([cx + r * Math.cos(angle), cy + r * Math.sin(angle)])
  }

  // Catmull-Rom spline for smooth curves
  function catmullRom(pts, steps = 12) {
    const result = []
    const n = pts.length
    for (let i = 0; i < n; i++) {
      const p0 = pts[(i - 1 + n) % n]
      const p1 = pts[i]
      const p2 = pts[(i + 1) % n]
      const p3 = pts[(i + 2) % n]
      for (let t = 0; t < steps; t++) {
        const tt = t / steps
        const t2 = tt * tt, t3 = t2 * tt
        result.push([
          0.5 * (2*p1[0] + (-p0[0]+p2[0])*tt + (2*p0[0]-5*p1[0]+4*p2[0]-p3[0])*t2 + (-p0[0]+3*p1[0]-3*p2[0]+p3[0])*t3),
          0.5 * (2*p1[1] + (-p0[1]+p2[1])*tt + (2*p0[1]-5*p1[1]+4*p2[1]-p3[1])*t2 + (-p0[1]+3*p1[1]-3*p2[1]+p3[1])*t3),
        ])
      }
    }
    return result
  }

  // Round toward circle for plumpness
  const smoothPts = catmullRom(rawPts, 16)
  const avgR = (outerR + innerR) / 2
  const roundFactor = 0.15
  const finalPts = smoothPts.map(([x, y]) => {
    const dx = x - cx, dy = y - cy
    const dist = Math.sqrt(dx * dx + dy * dy)
    if (dist === 0) return [x, y]
    const cx2 = cx + avgR * dx / dist
    const cy2 = cy + avgR * dy / dist
    return [x * (1 - roundFactor) + cx2 * roundFactor, y * (1 - roundFactor) + cy2 * roundFactor]
  })

  const starPath = `M ${finalPts[0][0]} ${finalPts[0][1]} ` +
    finalPts.slice(1).map(p => `L ${p[0]} ${p[1]}`).join(' ') + ' Z'

  // Slightly smaller star for highlight
  const hlPts = finalPts.map(([x, y]) => {
    const dx = x - cx, dy = y - cy
    return [cx + dx * 0.92, cy + dy * 0.92 - s * 0.008]
  })
  const hlPath = `M ${hlPts[0][0]} ${hlPts[0][1]} ` +
    hlPts.slice(1).map(p => `L ${p[0]} ${p[1]}`).join(' ') + ' Z'

  // Even smaller for inner glow
  const glowPts = finalPts.map(([x, y]) => {
    const dx = x - cx, dy = y - cy
    return [cx + dx * 0.75, cy + dy * 0.75 - s * 0.01]
  })
  const glowPath = `M ${glowPts[0][0]} ${glowPts[0][1]} ` +
    glowPts.slice(1).map(p => `L ${p[0]} ${p[1]}`).join(' ') + ' Z'

  // 4-point sparkle
  const sparkle = (sx, sy, sz, opacity = 0.85) => {
    const a = sz, b = sz * 0.28
    return `<path d="M ${sx} ${sy-a} C ${sx+b} ${sy-b}, ${sx+b} ${sy-b}, ${sx+a} ${sy} C ${sx+b} ${sy+b}, ${sx+b} ${sy+b}, ${sx} ${sy+a} C ${sx-b} ${sy+b}, ${sx-b} ${sy+b}, ${sx-a} ${sy} C ${sx-b} ${sy-b}, ${sx-b} ${sy-b}, ${sx} ${sy-a} Z" fill="white" opacity="${opacity}"/>`
  }

  return `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 ${s} ${s}" width="${s}" height="${s}">
  <defs>
    <!-- Rich background gradient -->
    <linearGradient id="bg1" x1="0" y1="0" x2="1" y2="1">
      <stop offset="0%" stop-color="#FFE082"/>
      <stop offset="35%" stop-color="#FFD54F"/>
      <stop offset="70%" stop-color="#FFB300"/>
      <stop offset="100%" stop-color="#FF9800"/>
    </linearGradient>

    <!-- Top-left light wash -->
    <radialGradient id="bgLight" cx="0.25" cy="0.2" r="0.7">
      <stop offset="0%" stop-color="white" stop-opacity="0.25"/>
      <stop offset="100%" stop-color="white" stop-opacity="0"/>
    </radialGradient>

    <!-- Bottom-right warm shadow -->
    <radialGradient id="bgDark" cx="0.8" cy="0.85" r="0.5">
      <stop offset="0%" stop-color="#E65100" stop-opacity="0.15"/>
      <stop offset="100%" stop-color="#E65100" stop-opacity="0"/>
    </radialGradient>

    <!-- Star body gradient - warm 3D -->
    <linearGradient id="starBody" x1="0.3" y1="0" x2="0.7" y2="1">
      <stop offset="0%" stop-color="#FFFDF5"/>
      <stop offset="40%" stop-color="#FFF9E6"/>
      <stop offset="100%" stop-color="#FFE9A0"/>
    </linearGradient>

    <!-- Star top highlight - glossy -->
    <linearGradient id="starHighlight" x1="0.5" y1="0" x2="0.5" y2="0.6">
      <stop offset="0%" stop-color="white" stop-opacity="0.9"/>
      <stop offset="50%" stop-color="white" stop-opacity="0.2"/>
      <stop offset="100%" stop-color="white" stop-opacity="0"/>
    </linearGradient>

    <!-- Star inner glow -->
    <radialGradient id="starInnerGlow" cx="0.45" cy="0.38" r="0.4">
      <stop offset="0%" stop-color="white" stop-opacity="0.7"/>
      <stop offset="100%" stop-color="white" stop-opacity="0"/>
    </radialGradient>

    <!-- Star rim light (edge highlight) -->
    <radialGradient id="rimLight" cx="0.3" cy="0.2" r="0.8">
      <stop offset="0%" stop-color="white" stop-opacity="0.4"/>
      <stop offset="50%" stop-color="white" stop-opacity="0.1"/>
      <stop offset="100%" stop-color="#FFB300" stop-opacity="0.05"/>
    </radialGradient>

    <!-- Shadow filter -->
    <filter id="starShadow" x="-15%" y="-10%" width="130%" height="140%">
      <feDropShadow dx="0" dy="${s*0.015}" stdDeviation="${s*0.025}" flood-color="#BF6C00" flood-opacity="0.35"/>
    </filter>

    <!-- Soft outer glow -->
    <filter id="outerGlow" x="-20%" y="-20%" width="140%" height="140%">
      <feGaussianBlur in="SourceGraphic" stdDeviation="${s*0.015}"/>
    </filter>

    <!-- Sparkle glow -->
    <filter id="sparkleGlow" x="-50%" y="-50%" width="200%" height="200%">
      <feGaussianBlur stdDeviation="${s*0.004}"/>
    </filter>

    <!-- Blush filter -->
    <filter id="blush" x="-50%" y="-50%" width="200%" height="200%">
      <feGaussianBlur stdDeviation="${s*0.015}"/>
    </filter>

    <!-- Eye shadow -->
    <filter id="eyeShadow">
      <feDropShadow dx="0" dy="${s*0.002}" stdDeviation="${s*0.003}" flood-color="#5D4037" flood-opacity="0.2"/>
    </filter>

    <!-- Clip for glossy overlay -->
    <clipPath id="starClip">
      <path d="${starPath}"/>
    </clipPath>
  </defs>

  <!-- === BACKGROUND === -->
  <rect width="${s}" height="${s}" rx="${cornerR}" ry="${cornerR}" fill="url(#bg1)"/>
  <rect width="${s}" height="${s}" rx="${cornerR}" ry="${cornerR}" fill="url(#bgLight)"/>
  <rect width="${s}" height="${s}" rx="${cornerR}" ry="${cornerR}" fill="url(#bgDark)"/>

  <!-- Subtle inner border for depth -->
  <rect x="${s*0.01}" y="${s*0.01}" width="${s*0.98}" height="${s*0.98}" rx="${Math.max(cornerR-s*0.01, 0)}" ry="${Math.max(cornerR-s*0.01, 0)}" fill="none" stroke="white" stroke-opacity="0.15" stroke-width="${s*0.004}"/>

  <!-- === STAR OUTER GLOW === -->
  <path d="${starPath}" fill="#FFF8E1" opacity="0.4" filter="url(#outerGlow)"/>

  <!-- === STAR WITH SHADOW === -->
  <g filter="url(#starShadow)">
    <!-- Star base -->
    <path d="${starPath}" fill="url(#starBody)"/>
  </g>

  <!-- Star edge/rim highlight -->
  <path d="${starPath}" fill="url(#rimLight)"/>

  <!-- Glossy overlay on top half -->
  <g clip-path="url(#starClip)">
    <ellipse cx="${cx}" cy="${cy - outerR * 0.35}" rx="${outerR * 0.85}" ry="${outerR * 0.55}" fill="url(#starHighlight)"/>
  </g>

  <!-- Inner glow center -->
  <path d="${glowPath}" fill="url(#starInnerGlow)" opacity="0.6"/>

  <!-- Subtle bottom edge shadow inside star -->
  <g clip-path="url(#starClip)">
    <ellipse cx="${cx}" cy="${cy + outerR * 0.7}" rx="${outerR * 0.6}" ry="${outerR * 0.2}" fill="#E8A000" opacity="0.08"/>
  </g>

  <!-- === CUTE FACE === -->
  <g filter="url(#eyeShadow)">
    <!-- Left eye -->
    <ellipse cx="${cx - outerR*0.21}" cy="${cy + outerR*0.01}" rx="${outerR*0.058}" ry="${outerR*0.068}" fill="#4E342E"/>
    <!-- Right eye -->
    <ellipse cx="${cx + outerR*0.21}" cy="${cy + outerR*0.01}" rx="${outerR*0.058}" ry="${outerR*0.068}" fill="#4E342E"/>
  </g>

  <!-- Eye highlights - big -->
  <ellipse cx="${cx - outerR*0.23}" cy="${cy - outerR*0.02}" rx="${outerR*0.028}" ry="${outerR*0.032}" fill="white"/>
  <ellipse cx="${cx + outerR*0.19}" cy="${cy - outerR*0.02}" rx="${outerR*0.028}" ry="${outerR*0.032}" fill="white"/>

  <!-- Eye highlights - small -->
  <circle cx="${cx - outerR*0.19}" cy="${cy + outerR*0.03}" r="${outerR*0.013}" fill="white" opacity="0.7"/>
  <circle cx="${cx + outerR*0.23}" cy="${cy + outerR*0.03}" r="${outerR*0.013}" fill="white" opacity="0.7"/>

  <!-- Blush -->
  <ellipse cx="${cx - outerR*0.35}" cy="${cy + outerR*0.10}" rx="${outerR*0.085}" ry="${outerR*0.055}" fill="#FF6E6E" opacity="0.3" filter="url(#blush)"/>
  <ellipse cx="${cx + outerR*0.35}" cy="${cy + outerR*0.10}" rx="${outerR*0.085}" ry="${outerR*0.055}" fill="#FF6E6E" opacity="0.3" filter="url(#blush)"/>

  <!-- Smile -->
  <path d="M ${cx - outerR*0.10} ${cy + outerR*0.10} Q ${cx} ${cy + outerR*0.20} ${cx + outerR*0.10} ${cy + outerR*0.10}"
    stroke="#5D4037" stroke-width="${outerR*0.025}" stroke-linecap="round" fill="none" opacity="0.85"/>

  <!-- === SPARKLES === -->
  <g filter="url(#sparkleGlow)">
    ${sparkle(cx + s*0.30, cy - s*0.24, s*0.028, 0.9)}
    ${sparkle(cx - s*0.33, cy - s*0.16, s*0.020, 0.75)}
    ${sparkle(cx + s*0.24, cy + s*0.30, s*0.017, 0.7)}
    ${sparkle(cx - s*0.27, cy + s*0.28, s*0.022, 0.65)}
    ${sparkle(cx + s*0.06, cy - s*0.37, s*0.014, 0.6)}
    ${sparkle(cx - s*0.15, cy + s*0.36, s*0.012, 0.5)}
  </g>

  <!-- Tiny round sparkle dots -->
  <circle cx="${cx + s*0.37}" cy="${cy - s*0.05}" r="${s*0.006}" fill="white" opacity="0.5"/>
  <circle cx="${cx - s*0.39}" cy="${cy + s*0.08}" r="${s*0.005}" fill="white" opacity="0.4"/>
  <circle cx="${cx + s*0.12}" cy="${cy + s*0.38}" r="${s*0.005}" fill="white" opacity="0.45"/>
  <circle cx="${cx - s*0.08}" cy="${cy - s*0.40}" r="${s*0.004}" fill="white" opacity="0.35"/>
  <circle cx="${cx + s*0.35}" cy="${cy + s*0.18}" r="${s*0.004}" fill="white" opacity="0.4"/>

</svg>`
}

async function generateHighQuality(targetSize, outputPath, maskable = false) {
  const renderSize = 1024
  const svg = createIconSvg(renderSize, maskable)
  await sharp(Buffer.from(svg))
    .resize(targetSize, targetSize, { kernel: 'lanczos3' })
    .png({ quality: 100 })
    .toFile(outputPath)
  console.log(`Generated: ${outputPath} (${targetSize}x${targetSize})`)
}

const base = 'public'
await generateHighQuality(180, `${base}/apple-touch-icon-180x180.png`)
await generateHighQuality(192, `${base}/pwa-192x192.png`)
await generateHighQuality(512, `${base}/pwa-512x512.png`)
await generateHighQuality(512, `${base}/maskable-icon-512x512.png`, true)

console.log('\nAll icons generated!')
